experimental/attestation_common/src/keying_material.rs

```diff
@@ -0,0 +1,146 @@
+            .context("Couldn't serialize keying material")?;
+        let tee_report = Report::new(&serialized_keying_material.as_bytes());
+        let attestation_info = AttestationInfo {
+            report: tee_report,
+            certificate: tee_certificate.to_vec(),
+        };
+
+        Ok(Self { attestation_info })
+    }
+
+    /// Verify remote attestation assertion by verifying the TEE report and checking that it
+    /// contains `expected_keying_material`.
+    pub fn verify(&self, expected_keying_material: &KeyingMaterial) -> anyhow::Result<()> {
+        self.attestation_info
+            .verify()
+            .context("Incorrect attestation info")?;
+
+        let serialized_expected_keying_material = serde_json::to_string(expected_keying_material)
+            .context("Couldn't serialize keying material")?;
+        if serialized_expected_keying_material.as_bytes() == self.attestation_info.report.data {
+            Ok(())
+        } else {
+            Err(anyhow!("Incorrect keying material"))
+        }
+    }
+
+    pub fn from_string(input: &str) -> anyhow::Result<Self> {
+        serde_json::from_str(input).context("Couldn't deserialize assertion")
+    }
+
+    pub fn to_string(&self) -> anyhow::Result<String> {
+        serde_json::to_string(&self).context("Couldn't serialize assertion")
+    }
+}
+
+/// Value uniquely derived from the TLS master secret that is shared by both participants of the TLS
+/// session. Used to bind remote attestation to each TLS session independently.
+/// https://tools.ietf.org/html/rfc5705
+#[derive(Clone, Serialize, Deserialize, PartialEq)]
+pub struct KeyingMaterial {
+    /// Value used to disambiguate the proper use of the `KeyingMaterial`
+    purpose: String,
+    /// Pseudo-random value that is uniquely exported for an individual TLS session.
+    value: Vec<u8>,
+}
+
+impl KeyingMaterial {
+    /// Export `KeyingMaterial` for a given TLS `session`.
+    /// https://tools.ietf.org/html/rfc5705#section-4
+    fn export(
+        // TLS session used to export `KeyingMaterial` from.
+        session: &dyn rustls::Session,
+        // Label value that is needed to distinguish between application level protocols that rely
+        // on `KeyingMaterial`.
+        label: &str,
+        // Context value that is used to derive separate `KeyingMaterial` for different assertion
+        // types.
+        assertion_type: &str,
+    ) -> anyhow::Result<Self> {
+        let mut buffer = vec![0; KEYING_MATERIAL_LENGTH];
+        session
+            .export_keying_material(
+                &mut buffer,
+                label.as_bytes(),
+                Some(assertion_type.as_bytes()),
+            )
+            .context("Couldn't export keying material")?;
+        use log::info;
+        info!(
+            "Server Keying material: {:?} (len: {})",
+            buffer,
+            buffer.len()
+        );
+        Ok(Self {
+            purpose: KEYING_MATERIAL_PURPOSE.to_string(),
```

Comment:
What's the point of this? Could you add a comment? And BTW, shouldn't we use a purpose string as prefix to the context value / assertion type?